x/programs/cmd/simulator/src/simulator.rs

```diff
@@ -0,0 +1,367 @@
+
+        self.writer.write_all(&input)?;
+        self.writer.write_all(b"'\n")?;
+        self.writer.flush()?;
+
+        self.responses
+            .next()
+            .ok_or(SimulatorError::Client(ClientError::Eof))?
+            .and_then(|resp: SimulatorResponse| {
+                if let Some(err) = resp.error {
+                    Err(SimulatorError::Program(err))
+                } else {
+                    Ok(resp)
+                }
+            })
+    }
+
+    pub fn read(&mut self, method: String, params: Vec<Param>) -> SimulatorResponseItem {
+        self.writer.write_all(Self::RUN_COMMAND)?;
+
+        let input = serde_json::to_vec(&SimulatorRequest::new_read(method, params))?;
+
+        self.writer.write_all(&input)?;
+        self.writer.write_all(b"'\n")?;
+        self.writer.flush()?;
+
+        self.responses
+            .next()
+            .ok_or(SimulatorError::Client(ClientError::Eof))?
+            .and_then(|request| {
+                if let Some(err) = request.error {
+                    Err(SimulatorError::Program(err))
+                } else {
+                    Ok(request)
+                }
+            })
+    }
+
+    pub fn execute(
+        &mut self,
+        method: String,
+        params: Vec<Param>,
+        max_units: u64,
+    ) -> SimulatorResponseItem {
+        self.writer.write_all(Self::RUN_COMMAND)?;
+
+        let input = serde_json::to_vec(&SimulatorRequest::new_execute(method, params, max_units))?;
+
+        self.writer.write_all(&input)?;
+        self.writer.write_all(b"'\n")?;
+        self.writer.flush()?;
+
+        self.responses
+            .next()
+            .ok_or(SimulatorError::Client(ClientError::Eof))?
+            .and_then(|request| {
+                if let Some(err) = request.error {
+                    Err(SimulatorError::Program(err))
+                } else {
+                    Ok(request)
+                }
+            })
+    }
+}
+
+/// A [`SimulatorRequest`] is a call to the simulator
+#[derive(Debug, Serialize, PartialEq, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct SimulatorRequest {
+    /// The API endpoint to call.
+    pub endpoint: Endpoint,
+    /// The method to call on the endpoint.
+    pub method: String,
+    /// The maximum number of units the Request can consume.
+    pub max_units: u64,
+    /// The parameters to pass to the method.
+    pub params: Vec<Param>,
+}
+
+impl SimulatorRequest {
+    pub fn new_read(method: String, params: Vec<Param>) -> Self {
+        Self {
+            endpoint: Endpoint::ReadOnly,
+            method: method.to_string(),
+            max_units: 0,
+            params,
+        }
+    }
+
+    pub fn new_execute(method: String, params: Vec<Param>, max_units: u64) -> Self {
+        Self {
+            endpoint: Endpoint::Execute,
+            method,
+            max_units,
+            params,
+        }
+    }
+
+    pub fn new_create_program(path: String) -> Self {
```

Comment:
The constructor is already implied by the fact that it's a static method. 
```suggestion
    pub fn read(method: String, params: Vec<Param>) -> Self {
        Self {
            endpoint: Endpoint::ReadOnly,
            method: method.to_string(),
            max_units: 0,
            params,
        }
    }

    pub fn execute(method: String, params: Vec<Param>, max_units: u64) -> Self {
        Self {
            endpoint: Endpoint::Execute,
            method,
            max_units,
            params,
        }
    }

    pub fn create_program(path: String) -> Self {
```